keps/sig-node/3960-pod-lifecycle-sleep-action/README.md

```diff
@@ -0,0 +1,436 @@
+  1. Create a simple pod with a container that runs a long-running process.
+  2. Add a preStop hook to the container configuration, using the new sleepAction with a specified sleep duration (e.g., 5 seconds).
+  3. Delete the pod and observe the time it takes for the container to terminate.
+  4. Verify that the container sleeps for the specified duration before it is terminated.
+  5. Verify that the container keeps executing code while sleeping.
+
+- Sleep duration boundary testing
+  1. Create a simple pod with a container that runs a long-running process.
+  2. Add a preStop hook to the container configuration, using the new sleepAction with various sleep durations, including:1 seconds (minimum allowed value), values slightly above the minimum allowed value (to test edge cases).
+  3. For each sleep duration, delete the pod and observe the time it takes for the container to terminate.
+  4. Verify that the container sleeps for the specified duration before it is terminated.
+  5. Verify that the container keeps executing code while sleeping.
+
+- Container exit/crash testing
+  1. Create a simple pod with a container that will exit soon.
+  2. Add a preStop hook to the container configuration, using the new sleepAction with a specified sleep duration longer than the container's lifecycle.
+  3. Exit the container and observe the time it takes for the pod to terminate.
+  4. Verify that the pod be deleted successfully without waiting for the entire sleep duration.
+
+- Interaction with termination grace period
+  1. Create a simple pod with a container that runs a long-running process.
+  2. Add a preStop hook to the container configuration, using the new sleepAction with a specified sleep duration (e.g., 5 seconds).
+  3. Set the termination grace period to various values, including:
+     - Equal to the sleep duration
+     - Greater than the sleep duration
+     - Greater than the sleep duration, but reduced to less than the sleep duration at runtime
+  4. For each termination grace period value, delete the pod and observe the time it takes for the container to terminate.
+  5. Verify that the container is terminated after the min(sleep, grace).
+
+### Graduation Criteria
+
+#### Alpha
+
+- Feature implemented behind a feature flag
+- Initial unit/e2e tests completed and enabled
+- Documentation is added to demonstrate why this is useful in nginx scenario and how exactly nginx needs to be configured.
+
+#### Beta
+
+- Gather feedback from developers and surveys
+- Additional e2e tests are completed(if needed)
+
+#### GA
+
+- No negative feedback
+- No bug issues reported
+
+### Upgrade / Downgrade Strategy
+
+#### Upgrade
+The previous PreStop behavior will not be broken. Users can continue to use their hooks as it is.
+To use this enhancement, users need to enable the feature gate, and add sleep action in their prestop hook.
+
+#### Downgrade
+Kube-apiserver will ignore sleep in prestop hook.
+
+### Version Skew Strategy
+
+N/A
```

Comment:
Disagree - it actually matters. It's definitely possible [even more, it will happen for sure at least for a moment] that the FG will be enabled only in one component and not the other.

Ideally, I would like to see  matrix of:
- kubeapiserver enabled/disable
- kubelet enabled/disable
with description of what exactly we expect

Reply:
Updated, but I'm not sure what will happen in the scenario `only the kube-apiserver enable this feature`, will the creating request pass the validation and successsfully processed or will it be rejected?

Reply:
The semantic that I believe works best is close to (2) [although I didn't fully understand "and nullifies ..." part]
I believe what we should target is:

If FG is disabled:
(a) any attempt to set the field for an object doesn't succeed - the field is silently dropped - the "dropFields" strategy: https://github.com/kubernetes/kubernetes/blob/master/pkg/registry/core/pod/strategy.go#LL89C10-L89C31
(b) if the field was set once the FG was enabled, it stays to be set
[the above is what kube-apiserver is doing]

(c) if the FG is disabled, even if Kubelet (or in general any other component) observes the field it ignores its existence

[for (c) there are exceptions - the nice example appeared in Sidecar KEP and computing resources in scheduler, but I personally treat them as exceptions - by default I consider the above the desired semantic]

@thockin - do you have any concerns about this?

Reply:
For this specific field, that seems somewhat acceptable because it only matters exactly once (when pod is deleted), although it is pretty surprising that the API says it is on but its really not.

But imagine a field which is used in real-time on a long-lived resource like Service or Deployment.  The object was admitted, the feature was used, then the gate was disabled, the API stil says the feature is on, but suddenly it stopped working.  Worse - it could stop working on some components and not others (e.g. disabled on some kube-proxy and noth others).

That makes testing MUCH more complicated.

Now imagine an enum sort of field or a loosening of validation, where a new value was allowed by the gate, and then the gate is disabled, and the value is no longer allowable.  Do we fall back on some next-closest value without updating the API?  That seems terrible.

Reply:
As for the real-time example - I would actually say that the "stops working" (or "partially stops working") is still what I want. If I'm disabling a FG, I'm doing it on purpose, so I actually want this feature to really stop working, as this is presumably causing some troubles.
I definitely agree that the current support for it is poor, but I still think it's what I want.
And eventually the FeatureGate KEP will address the issue [you register a hook for disabling FG that is clearing this field, or doing whatever else is better in a given case]

Re enum example - this is harder one. The only ones that I've seen where effectively "yet another type of optional behavior", so disabling meant "you no longer have this, so you don't get anything". Which is kind of the same case as the above.
If it won't be an "optional behavior", but rather "some behavior is required" that would become super tricky and that may justify an exception. But I guess it depends on the specific case.


So I guess the summary of my answer is: if I'm requesting disabling a FG, I actually expect the feature to become disabled. And I acknowledge the fact that if someone was using it, they may got affected/broken, but I'm disabling FG for a reason.

@deads2k @johnbelamaric - FYI - this is an interesting discussion

Reply:
@jpbetz also :)

Reply:
So, in this kep, should the kubelet have the gate?

if yes:
1.  only apiserver enable the FG
-  validation will pass, but kubelet will ignore this field when pod is terminating?

2.  only kubelet enable the FG
-   new pod will fail the validation, but the existing pods will exec the sleepAction.

if no:
- FG only controls the validation, once a container with sleepAction is set, it will always sleep before terminating regardless of whether FG still holds

Am I understanding this correct?

Reply:
THis is correct.

And what I'm saying is that we should go with the first option.
I know it has drawbacks, but I think the gains from it outweigh those drawbacks. It seems that Tim is far from being convinced on the generic case, but he seems to be ok for this particular case, so let's go for it here.

Reply:
ACK